### PR TITLE
Typescript Definitions

### PR DESCRIPTION
**Pull request by clouless**
Hi @iensu

I like you tool but want to use it with TypeScript.
This pull request bundles the typescript definitions together with the code.

cheers,
Bernhard

### CODE REVIEW

**Comment by iensu** [comment]
Hi Bernhard! 

Thank you for your PR. I am very sorry for the late response. I will have a proper look at this PR as soon as I can.

**Comment by clouless** [comment]
hi jens,

no problem :) I already totally forgot the PR too :D 
so take you time :)

File: package.json [hunk 1]
```diff
@@ -3,6 +3,7 @@
   "version": "3.0.0",
   "description": "A Mocha BDD integration",
   "main": "mocha-cakes.js",
+  "typings": "mocha-cakes",
```
Comment (iensu) [inline package.json]: This should probably be `"typings": "./mocha-cakes.d.ts"` (see http://www.typescriptlang.org/docs/handbook/declaration-files/publishing.html)

Comment (clouless) [reply package.json]: i fixed that to `"typings": "./mocha-cakes.d.ts"`

**Comment by clouless** [comment]
@iensu I am not so good when it comes to squashing. Could you not just squash on Merge?

**Comment by iensu** [comment]
Hi again! 

That did not work. I now see that `mocha-cakes.d.ts` is not listed in the `files` field in `package.json` which means that it is not being included in the package. Could you please add that as well?

**Comment by clouless** [comment]
@iensu ok :) included in package.json files array => https://github.com/clouless/mocha-cakes-2/commit/517c43b0b6e8c292474907f1609a7b5b4b018559

**Comment by iensu** [comment]
Now the types are accessible! However, I noticed that the property functions `.only` and `.skip` are not taken into account. I rewrote the definitions file like this, not sure how idiomatic it is though:

```typescript
interface MochaCakesTestFn {
  (t: string, f: Function): any;
  only(t: string, f: Function): any;
  skip(t: string, f: Function): any;
}

declare var Feature: MochaCakesTestFn;
declare var Scenario: MochaCakesTestFn;
declare var Given: MochaCakesTestFn;
declare var And: MochaCakesTestFn;
declare var When: MochaCakesTestFn;
declare var Then: MochaCakesTestFn;
declare var But: MochaCakesTestFn;

declare var feature: MochaCakesTestFn;
declare var scenario: MochaCakesTestFn;
declare var given: MochaCakesTestFn;
declare var and: MochaCakesTestFn;
declare var when: MochaCakesTestFn;
declare var then: MochaCakesTestFn;
declare var but: MochaCakesTestFn;
```

What do you think?

File: README.md [hunk 2]
```diff
@@ -117,6 +117,42 @@ Feature('Some feature', () => {
 
 <img src="doc/only-v2.png" width="400" />
 
+
+### TypeScript
+
+The TypeScript definitions are bundled together with mocha-cakes-2.
+When using TypeScript run mocha with [`mocha -r ts-node/register ...`](https://github.com/TypeStrong/ts-node).
+
```
Comment (iensu) [inline README.md]: Moght be good to add something about needing to install @types/mocha as well, since most of the utility functions are declared there.

**Comment by iensu** [comment]
Ok, I see. Thank you again for your time.

That's weird, for me it is working. 

Btw, I think we can have a look at the Mocha types definition for inspiration: https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/mocha/index.d.ts
It's a bit complex though.

**Comment by clouless** [comment]
@iensu I think it highly depends on the tsconfig.json you use. Could you paste your `tsconfig.json` here thx. I will setup a test project with typescript where we can test our stuff. Then if we both say it works, we can merge :)

**Comment by iensu** [comment]
I have no tsconfig file in my repo. Will try your test-repo!

**Comment by iensu** [comment]
I've tested your sandbox project and for me it works. The only compilation error I got was that `expect` was not definied, but I fixed that by explicitly importing it from chai. 

I also got some warnings about the `result`, `a` and `b` variable types, but got rid of those by specifying the number type (`let result: number` etc). 

My typescript version is 2.7.2

**Comment by clouless** [comment]
very strange, I get "Feature is not a function" on the test project. Maybe it depends on nodejs and other versions .... I will have a look tonight after work :)

**Comment by iensu** [comment]
Very strange indeed :) I'll see if I can convince some colleagues to try it out here aswell and see if it works for them

**Comment by clouless** [comment]
What happens when you do:
Uninstall global mocha, global typescript, global ts-node:

```
npm uninstall -g mocha
npm uninstall -g typescript
npm uninstall -g ts-node
```

And then do:

```
git clone https://github.com/clouless/mocha-cakes-2-typescript-sandbox.git
cd mocha-cakes-2-typescript-sandbox
yarn install

yarn test
```

I tested it on macOS and with official docker image:

```
docker run -i -t node:9.7.1 bash
### INSIDE DOCKER
git clone https://github.com/clouless/mocha-cakes-2-typescript-sandbox.git
cd mocha-cakes-2-typescript-sandbox

#### update
apt-get update && apt-get install apt-transport-https

#### install yarn
curl -sS https://dl.yarnpkg.com/debian/pubkey.gpg | apt-key add -
echo "deb https://dl.yarnpkg.com/debian/ stable main" | tee /etc/apt/sources.list.d/yarn.list
apt-get update && apt-get install --no-install-recommends yarn

#### Install packages
yarn install

#### run tests
yarn test
```

I get exactly what I get on macOS

image

**Comment by iensu** [comment]
Ha! Now I'm getting the same errors! I wonder how those globals ended up there, I'm usually very careful never to install things globally.

**Comment by iensu** [comment]
Change your test script to: `mocha -r ts-node/register --ui mocha-cakes-2 src/*.spec.ts`.

You are not loading the mocha-cakes-2 plugin into mocha, so it cannot find the functions

**Comment by clouless** [comment]
oh my. You are right, I will try it out today, and then we should be ok hopefully :)

**Comment by iensu** [comment]
No problem, we've all been there :)

Thanks for taking your time and for your patience with the PR :)

**Pull request merged**